pkg/operator/configobservation/images/observe_images.go

```diff
@@ -56,3 +59,117 @@ func ObserveInternalRegistryHostname(genericListers configobserver.Listers, reco
 	}
 	return observedConfig, errs
 }
+
+func ObserveExternalRegistryHostnames(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	var errs []error
+	prevObservedConfig := map[string]interface{}{}
+
+	// first observe all the existing config values so that if we get any errors
+	// we can at least return those.
+	externalRegistryHostnamePath := []string{"imagePolicyConfig", "externalRegistryHostnames"}
+	o, _, _ := unstructured.NestedSlice(existingConfig, externalRegistryHostnamePath...)
+	if len(o) > 0 {
+		err := unstructured.SetNestedSlice(prevObservedConfig, o, externalRegistryHostnamePath...)
+		if err != nil {
+			return prevObservedConfig, append(errs, err)
+		}
+	}
+
+	if !listers.ImageConfigSynced() {
+		glog.Warning("images.config.openshift.io not synced")
+		return prevObservedConfig, errs
+	}
+
+	// now gather the cluster config and turn it into the observed config
+	observedConfig := map[string]interface{}{}
+	configImage, err := listers.ImageConfigLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		glog.Warningf("image.config.openshift.io/cluster: not found")
+		return observedConfig, errs
+	}
+	if err != nil {
+		return prevObservedConfig, append(errs, err)
+	}
+
+	// User provided values take precedence, first entry in the array
+	// has special significance.
+	externalRegistryHostnames := configImage.Spec.ExternalRegistryHostnames
+	externalRegistryHostnames = append(externalRegistryHostnames, configImage.Status.ExternalRegistryHostnames...)
+
+	if len(externalRegistryHostnames) > 0 {
```

Comment:
what happens when somebody decide to set this field to `""` because they don't want any external hostname for their registry anymore?

Reply:
We return an observedconfig that doesn't contain the field.  So i guess you tell me what happens at that point?  I would have assumed it would result in a new config that doesn't contain any setting for externalRegistryHostname.


If you guys aren't throwing away the existing config, then i don't know why we have to jump through all these "fetch the previous config an ensure we return it if an error happens" hoops.

Reply:
The resulting config does have the empty string because the default config has that value: https://github.com/openshift/cluster-kube-apiserver-operator/blob/819b12320e7f85065fa49f77f387900ab0367b24/bindata/v3.11.0/kube-apiserver/defaultconfig.yaml#L59

If you don't return a field in the observed config here, the merging algorithm does not change the defaultconfig.yaml value. So I think here it is ok, to do nothing in the observer. If there was a non-empty value in the default config, we would have to interpret an empty value by the user in a different way.